openlabcmd/openlabcmd/zk.py

```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
         return sorted(result)
 
-    @_client_check_wrapper
-    def get_service(self, service_name, role=None):
-        if not role:
-            if (service_name in
-                    service.service_mapping['master_service']['necessary'] +
-                    service.service_mapping['master_service']['unnecessary']):
-                role = 'master'
-            elif (service_name in
-                  service.service_mapping['slave_service']['necessary'] +
-                  service.service_mapping['slave_service']['unnecessary']):
-                role = 'slave'
-            else:
-                exceptions.ClientError("Can't find service %s" % service_name)
+    def _get_service_role(self, service_name):
+        if (service_name in
+                service.service_mapping['master_service']['necessary'] +
+                service.service_mapping['master_service']['unnecessary']):
+            return 'master'
+        elif (service_name in
+              service.service_mapping['slave_service']['necessary'] +
+              service.service_mapping['slave_service']['unnecessary']):
+            return 'slave'
 
+        exceptions.ClientError("Can't find service %s" % service_name)
+
+    def _get_service_path_and_node(self, service_name, role):
         for exist_node in self.list_nodes():
             if exist_node['role'] == role:
                 path = '/ha/%s/%s_service/%s' % (exist_node['name'], role,
                                                  service_name)
-                result = self.client.get(path)
-                result = self._bytesToDict(result[0])
-                return result
+                return path, exist_node
         raise exceptions.ClientError("Can't find service %s" % service_name)
+
+    @_client_check_wrapper
+    def get_service(self, service_name, role=None):
+        if not role:
+            role = self._get_service_role(service_name)
+        path, srv_node = self._get_service_path_and_node(service_name, role)
+        result = self.client.get(path)
+        mtime = result[1].mtime
+        result = self._bytesToDict(result[0])
+        result['updated_at'] = datetime.datetime.fromtimestamp(
+            mtime / 1000).isoformat()
+        result['node_name'] = srv_node['name']
+        return result
+
+    @_client_check_wrapper
+    def update_service(self, service_name, role=None, alarmed=None,
+                       restarted=None, status=None, **kwargs):
+        old_service = self.get_service(service_name, role)
+        if not role:
+            if service_name in service.MIXED_SERVICE:
+                raise exceptions.ValidationError(
+                    "Role must be specified if service is in "
+                    "%s." % service.MIXED_SERVICE)
+            role = self._get_service_role(service_name)
+        path, _ = self._get_service_path_and_node(service_name, role)
+        current_time = datetime.datetime.now().isoformat()
+
+        if alarmed:
```

Comment:
How to update `alarmed` to false?

Reply:
oh, my fault here. It should be `if alarmed is not None`